### PR TITLE
MNT-97: Add support for passing product option id, title and price

### PR DESCRIPTION
**Pull request by radkomih**
Sub-story of MNT-94/Pass product url and image url with orders from the backend

https://metrilojira.atlassian.net/browse/MNT-97

### CODE REVIEW

File: lib/metracker/types/product.rb [hunk 1]
```diff
@@ -5,21 +5,28 @@ module Types
     class Product < Base
       attr_reader :id, :name, :price, :url, :image_url
 
-      def initialize(id:, price:, name:, url:, image_url:)
-        @id        = id
-        @name      = name
-        @price     = price
-        @url       = url
-        @image_url = image_url
+      def initialize(id:, price:, name:, url:, image_url:,
```
Comment (yassenb) [inline lib/metracker/types/product.rb]: When you have such an absurdly long list of parameters to a function it's clear something is wrong. I don't know what "option" is in this case but it seems it should be a separate object

File: spec/metracker/api_calls/add_to_cart_spec.rb [hunk 2]
```diff
@@ -34,12 +37,15 @@
         'uid'             => session_uid,
         'event_type'      => 'add_to_cart',
         'event_params'    => {
-          'id'        => '553399',
+          'id'        => '555111',
```
Comment (yassenb) [inline spec/metracker/api_calls/add_to_cart_spec.rb]: This hash is basically copy-pasted from the above one. Extract to a common variable

File: spec/metracker/api_calls/order_spec.rb [hunk 3]
```diff
@@ -15,11 +15,14 @@
     end
 
     let(:product) do
-      Metracker::Types::Product.new(id: '553399',
+      Metracker::Types::Product.new(id: '555111',
```
Comment (yassenb) [inline spec/metracker/api_calls/order_spec.rb]: What are these ID-s? Why did you change them when you updated the spec?

File: spec/metracker/api_calls/view_product_spec.rb [hunk 4]
```diff
@@ -42,13 +45,16 @@
         'uid'             => session_uid,
         'event_type'      => 'view_product',
         'event_params'    => {
-          'id' => '553399',
+          'id' => '555111',
           'name' => 'Kinoa',
           'price' => 12.98,
-          'url' => 'https://shopshop.com/products/kinoa?id=553399',
-          'image_url' => 'https://cdn.shopshop.com/products/kinoa?id=553399',
+          'url' => 'https://shopshop.com/products/kinoa?variant=2233',
```
Comment (yassenb) [inline spec/metracker/api_calls/view_product_spec.rb]: Why does one change introduce a change that's basically a copy-paste in 5 tests? There's obviously something wrong with tests. Don't copy-paste

File: lib/metracker/types/product.rb (hunk 1 above)
Comment (radkomih) [reply lib/metracker/types/product.rb]: I meant to be a quick change :) But I'll move all option params into separate object since all  are related to a product variant.

File: spec/metracker/api_calls/add_to_cart_spec.rb (hunk 2 above)
Comment (radkomih) [reply spec/metracker/api_calls/add_to_cart_spec.rb]: All these "ID" changes aren't really needed for the sake of the specs, I just did it to better differentiate main product ids from product variant ids

File: spec/metracker/api_calls/view_product_spec.rb (hunk 4 above)
Comment (radkomih) [reply spec/metracker/api_calls/view_product_spec.rb]: ^

File: spec/metracker/api_calls/order_spec.rb (hunk 3 above)
Comment (radkomih) [reply spec/metracker/api_calls/order_spec.rb]: ^

File: lib/metracker/types/product.rb [hunk 5]
```diff
@@ -5,21 +5,28 @@ module Types
...
-          'url'       => @url,
-          'image_url' => @image_url
+          'id'           => @id,
+          'name'         => @name,
+          'price'        => @price,
+          'url'          => @url,
+          'image_url'    => @image_url,
+          'option_id'    => @option_id,
+          'option_name'  => @option_name,
+          'option_price' => @option_price
```
Comment (evgeni-spasov-metrilo) [inline lib/metracker/types/product.rb]: There are attribute readers for those variables, you can use them instead.
You don't have attribute readers for `@option_..` .

File: spec/metracker/api_calls/view_product_spec.rb (hunk 4 above)
Comment (yassenb) [reply spec/metracker/api_calls/view_product_spec.rb]: How is this related to the above?

File: spec/metracker/api_calls/add_to_cart_spec.rb (hunk 2 above)
Comment (yassenb) [reply spec/metracker/api_calls/add_to_cart_spec.rb]: Still that doesn't address the copy-paste elephant in the room

**Comment by radkomih** [comment]
I've removed the duplication.

File: spec/metracker/api_calls/remove_from_cart_spec.rb [hunk 6]
```diff
@@ -34,12 +41,15 @@
...
         'event_type'      => 'remove_from_cart',
         'event_params'    => {
-          'id'        => '553399',
+          'id'        => '621147',
           'name'      => 'Kinoa',
           'price'     => 12.98,
-          'url'       => 'https://shopshop.com/products/kinoa?id=553399',
-          'image_url' => 'https://cdn.shopshop.com/products/kinoa?id=553399',
-          'quantity'  => 1
+          'url'       => 'https://shopshop.com/products/kinoa?variant=7622',
```
Comment (yassenb) [inline spec/metracker/api_calls/remove_from_cart_spec.rb]: Duplication is still here. 4 times

File: spec/metracker/api_calls/view_product_spec.rb [hunk 7]
```diff
@@ -7,53 +7,77 @@
...
-          'name' => 'Kinoa',
-          'price' => 12.98,
-          'url' => 'https://shopshop.com/products/kinoa?id=553399',
-          'image_url' => 'https://cdn.shopshop.com/products/kinoa?id=553399',
+          'id'           => product_id,
+          'name'         => product_name,
+          'price'        => product_price,
+          'url'          => product_url,
+          'image_url'    => product_image_url,
+          'option_id'    => variant_id,
```
Comment (yassenb) [inline spec/metracker/api_calls/view_product_spec.rb]: Extracting things to `let`-s doesn't make it any less of a copy-paste

**Comment by radkomih** [comment]
Now should be ok

File: lib/metracker/types/product.rb [hunk 8]
```diff
@@ -3,24 +3,27 @@
 module Metracker
   module Types
     class Product < Base
-      attr_reader :id, :name, :price, :url, :image_url
+      attr_reader :id, :name, :price, :url, :image_url, :variant
```
Comment (yassenb) [inline lib/metracker/types/product.rb]: What's the use of these `attr_reader`-s everywhere? They're not used and they only lead to more code to maintain

File: spec/metracker/support/shared_contexts/api_call.rb [hunk 9]
```diff
@@ -0,0 +1,69 @@
+shared_context 'base_api_call' do
+	let(:project_token) { 'project_token_moken' }
```
Comment (yassenb) [inline spec/metracker/support/shared_contexts/api_call.rb]: Indentation is all over the place. Configure your editor to never insert tabs, only spaces

Comment (radkomih) [reply spec/metracker/support/shared_contexts/api_call.rb]: Switched to sublime and forgot to configure that

File: lib/metracker/types/product.rb (hunk 8 above)
Comment (radkomih) [reply lib/metracker/types/product.rb]: Actually I use some of them inside the class and to be able to access the attributes when calling from different object.
All the rest I've declared to be consistent.

**Comment by adamof** [comment]
Whats up with this?

**Comment by radkomih** [comment]
It's ready for review

**Comment by adamof** [comment]
@yassenb can you have a look :)

**Comment by yassenb** [comment]
I don't see a commit addressing only the latest comments so I won't review again
